### PR TITLE
fix(app-cmds): incomplete payloads giving tracebacks within autocomplete

### PR DESCRIPTION
**Pull request by Skelmis**
## Summary

When using either a `nextcord.User` or `nextcord.Member` item within autocomplete there is a chance that the required data is missing due to discord not sending the data. This PR fixes the raw tracebacks and instead takes a best effort approach to attempting to get a valid object from cache before returning atleast a `nextcord.Object` object.

#### Reproduction code

```py

@client.slash_command()
async def test(
    interaction: nextcord.Interaction,
    member: nextcord.Member = nextcord.SlashOption(
        name="member", description="Select member", required=True
    ),
    warn: int = nextcord.SlashOption(name="warn", description="Warn", required=True),
):
    print(type(member))


@test.on_autocomplete("warn")
async def warn_autocomplete(interaction: nextcord.Interaction, warn: int, member: nextcord.Member):
    print(type(member))

    return await interaction.response.send_autocomplete([str(i) for i in range(10)])
```

#### Reproduction steps

Run the command, when populating the `member` choices click a user and do nothing else. Then click and fill in the `warn` entry, this should produce a traceback

## Checklist

- [x] If code changes were made then they have been tested.
  - [x] I have updated the documentation to reflect the changes.
  - [x] I have run `task pyright` and fixed the relevant issues.
- [x] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: nextcord/application_command.py [hunk 1]
```diff
@@ -1659,7 +1664,28 @@ async def handle_value(
...
             user_dict = {user.id: user for user in get_users_from_interaction(state, interaction)}
-            value = user_dict[user_id]
+            try:
+                value = user_dict[user_id]
+            except KeyError:
+                # By here the interaction data doesn't contain
+                # a full member/user object yet so fall back to bot cache
+                value = None
+                data = interaction.data
+                data = cast(ApplicationCommandInteractionData, data)
```
Comment (teaishealthy) [inline nextcord/application_command.py:1674]:
```suggestion
                data = cast(ApplicationCommandInteractionData, interaction.data)
```

File: nextcord/application_command.py [hunk 2]
```diff
@@ -1659,7 +1664,28 @@ async def handle_value(
...
+                guild_id = data.get("guild_id")
+                if guild_id:
+                    guild = state._guilds.get(int(guild_id))
+                    if guild:
+                        value = guild.get_member(user_id)
```
Comment (teaishealthy) [inline nextcord/application_command.py:1679]:
```suggestion
                if guild_id := data.get("guild_id"):
                    if guild := state._guilds.get(int(guild_id)):
                        value = guild.get_member(user_id)
```

File: nextcord/application_command.py [hunk 3]
```diff
@@ -1189,6 +1190,10 @@ async def favorite_dog(interaction: Interaction, dog: str):
             To use inputs from other options inputted in the command, you can add them as arguments to the autocomplete
             callback. The order of the arguments does not matter, but the names do.
 
+            If you are using :py:class:`nextcord.Member` or :py:class:`nextcord.user` typehints in
+            your autocompletes then just note that the objects provided can be both `None` and
+            :py:class:`nextcord.Object` depending on the data discord sends.
```
Comment (teaishealthy) [inline nextcord/application_command.py:1195]:
```suggestion
            :py:class:`nextcord.Object` depending on the data Discord sends.
```

File: nextcord/application_command.py [hunk 4]
```diff
@@ -1189,6 +1190,10 @@ async def favorite_dog(interaction: Interaction, dog: str):
             To use inputs from other options inputted in the command, you can add them as arguments to the autocomplete
             callback. The order of the arguments does not matter, but the names do.
 
+            If you are using :py:class:`nextcord.Member` or :py:class:`nextcord.user` typehints in
```
Comment (teaishealthy) [inline nextcord/application_command.py:1193]:
```suggestion
            If you are using :py:class:`nextcord.Member` or :py:class:`nextcord.User` typehints in
```

**Review by teaishealthy — CHANGES_REQUESTED** [review]

File: nextcord/application_command.py (hunk 4 above)
Comment (ooliver1) [inline nextcord/application_command.py:1193]:
```suggestion
            If you are using :class:`nextcord.Member` or :class:`nextcord.user` typehints in
```
Sphinx knows it is Python

File: nextcord/application_command.py (hunk 3 above)
Comment (ooliver1) [inline nextcord/application_command.py:1195]:
```suggestion
            :class:`nextcord.Object` depending on the data discord sends.
```

**Review by ooliver1 — CHANGES_REQUESTED** [review]

**Review by alentoghostflame — APPROVED** [review]
I haven't tested it myself, but I assume normal command usage and not JUST autocomplete has been tested already.
LGTM.

File: nextcord/application_command.py [hunk 5]
```diff
@@ -1659,7 +1664,25 @@ async def handle_value(
...
+                        value = guild.get_member(user_id)
+
+                if value is None:
+                    # Either we aren't in a guild or
+                    # the member object is not cached
+                    value = state._users.get(user_id)
+
+                if value is None:
+                    # Fall back to a Object at-least
+                    value = Object(id=user_id)
```
Comment (teaishealthy) [inline nextcord/application_command.py:1685]:
```suggestion
                    value = state._users.get(user_id)

                    if value is None:
                        # Fall back to a Object at-least
                        value = Object(id=user_id)
```

Check unnecessary if first if was False

**Review by teaishealthy — CHANGES_REQUESTED** [review]

**Review by teaishealthy — APPROVED** [review]

**Comment by Skelmis** [comment]
Ah well, I can't figure out how to dismiss reviews

**Review by ooliver1 — APPROVED** [review]

**Pull request merged**